### PR TITLE
separate index data into index.db

### PR DESCRIPTION
**Pull request by dustinxie**
this PR has 3 commits:

1. commit 1 defines interface for the block/action Indexer, added new code in /blockindex
2. commit 2 refactors indexing-related code in blockdao.go and blockchain.go using new interface
3. commit 3 improve the height to file name index. Instead of storing filename index for each individual block height, it only stores heights where new files are created into a range index

test plan:
make test
make minicluster
also i am running full-node sync test

commit 4: address comments per discussion with Zhefeng
commit 5: pass genesis hash to indexer

### CODE REVIEW

File: blockchain/blockdao/blockdao.go [hunk 1]
```diff
@@ -79,32 +71,31 @@ var (
...
+	GetFileIndex(uint64) ([]byte, error)
 	KVStore() db.KVStore
 }
 
 type blockDAO struct {
 	writeIndex    bool
 	compressBlock bool
 	kvstore       db.KVStore
+	indexer       blockindex.Indexer
+	htf           db.RangeIndex
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/blockdao/blockdao.go]: height to file index

File: blockindex/indexer.go [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+	// blockIndexer implements the Indexer interface
+	blockIndexer struct {
+		mutex       sync.RWMutex
+		genesisHash hash.Hash256
+		kvstore     db.KVStore
+		batch       db.KVStoreBatch
+		dirtyAddr   addrIndex
+		tbk         db.CountingIndex
+		tac         db.CountingIndex
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/indexer.go]: total block, total action

File: blockindex/indexer.go [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+// commit() writes the changes
+func (x *blockIndexer) commit() error {
+	var commitErr error
+	for k, v := range x.dirtyAddr {
+		if commitErr == nil {
+			if err := v.Commit(); err != nil {
+				commitErr = err
+			}
+		}
+		delete(x.dirtyAddr, k)
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/indexer.go]: i think so, say if the actions involve 5 accounts/dirtyaddr, and committing 2nd account somehow failed, we should purge all 5 account to restore x.dirtyAddr back to clean slate

File: blockindex/indexer.go [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+	totalActionsBucket = []byte("ac")
+	// ErrActionIndexNA indicates action index is not supported
+	ErrActionIndexNA = errors.New("action index not supported")
+)
+
+type (
+	addrIndex map[hash.Hash160]db.CountingIndex
+
+	// Indexer is the interface for block indexer
+	Indexer interface {
```
Comment (CoderZhi) [inline blockindex/indexer.go]: we can move the definition into blockchain/
Indexer interface could be as simple as
```
    Indexer interface {
      Start(context.Context) error
      Stop(context.Context) error
      PutBlock(*block.Block) error
      DeleteBlock(*block.Block) error
      Commit() error
    }
```

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -99,30 +100,18 @@ type Blockchain interface {
...
-	// GetBlockHashByActionHash returns Block hash by action hash
-	GetBlockHashByActionHash(h hash.Hash256) (hash.Hash256, error)
 	// GetReceiptsByHeight returns action receipts by block height
 	GetReceiptsByHeight(height uint64) ([]*action.Receipt, error)
 	// GetFactory returns the state factory
 	GetFactory() factory.Factory
 	// GetBlockDAO returns the block DAO
 	GetBlockDAO() blockdao.BlockDAO
+	// GetIndexer returns the chain indexer
+	GetIndexer() blockindex.Indexer
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: this is duplicated with GetBlockDao. We may merge them into one as
```Indexer() blockchain.Indexer```

File: blockchain/blockchain.go [hunk 6]
```diff
@@ -172,6 +161,7 @@ type Blockchain interface {
 type blockchain struct {
 	mu            sync.RWMutex // mutex to protect utk, tipHeight and tipHash
 	dao           blockdao.BlockDAO
+	indexer       blockindex.Indexer
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: correspondingly, we can delete it, and only keep dao

File: blockchain/blockchain.go (hunk 5 above)
Comment (CoderZhi) [reply blockchain/blockchain.go]: After rethinking about it, the API of blockchain may not need to have interfaces like GetFactory and GetBlockDao. Our definition of Blockchain is a mix of Blockchain and BlockchainBuilder, that's why we can StateFactoryOption, BlockDaoOption and etc.

**Pull request merged**